### DISCUSSION TITLE
Problem in writing application test

### DISCUSSION THREAD

**Discussion by mazarkaman** [abpframework/abp#14258] [Q&A]
Hi,

Imagine I have two application services:
`WebInfoAppService ` and `MainAppService` 

MainAppService is:

```
public class MainAppService:ApplicationService, IMainAppService
{
    private readonly IWebInfoAppService _webInfoAppService;

    public MainAppService(IWebInfoAppService webInfoAppService)
    {
        _webInfoAppService = webInfoAppService;
    }
    [Authorize]
    public async Task<IdentityUserDto> GetUserIfConnectedAsync()
    {
        if (await _webInfoAppService.IsConnectedAsync())
        {
            return new IdentityUserDto()
            {
                Id = CurrentUser.Id.Value,
                Email = CurrentUser.Email
            };
        }
        throw new UserFriendlyException("Not Connected");
    }
}
```

now I want to write a test for this class:

```
public class TestMainAppService:TestTestApplicationTestBase
{
    private readonly IMainAppService _mainService;

    public TestMainAppService()
    {
        var webInfoService = Substitute.For<IWebInfoAppService>();
        webInfoService.IsConnectedAsync().Returns(Task.FromResult(true));
        
       _mainService = new MainAppService(webInfoService);
    }

    [Fact]
    public async Task TestGetUserIfConnectedAsync()
    {
        var user = await _mainService.GetUserIfConnectedAsync();

        user.Id.ShouldNotBe(Guid.Empty);
    }
}
```

I should use Substitute.For in order to set return value for `IsConnectedAsync ` method of `IWebInfoAppService` class so i have to declare `MainAppService` Without using `GetRequiredService<IMainAppService>()` the problem is `**CurrentUser**` always null and if i use `GetRequiredService<IMainAppService>()` , CurrentUser will filled with data. what should I do to either change class behavior after it's injected or get data in CurrentUser if I declare WebInfoAppService directly?

**Comment by maliming** [comment]
hi

https://github.com/abpframework/abp/blob/22bc56e2794f0f323084df1ea680a1d1bdcded3c/templates/app/aspnet-core/test/MyCompanyName.MyProjectName.TestBase/Security/FakeCurrentPrincipalAccessor.cs

**Reply by mazarkaman** [reply]
dear @maliming I'm using this approach, but it works only when you declare the service by `GetRequiredService<IMainAppService>()` not when you declare it with the `new` keyword as I mentioned in the example. I need to declare my service directly because I want to change the behavior of injected service by NSubstitute.

**Comment by maliming** [comment]
> not work when you declare it with the new keyword 

`Substitute.For` creates a fake object that overrides all methods, not a single method.

You can set the ` webInfoService.GetUserIfConnectedAsync().Returns()`.

`FakeCurrent` user info from https://github.com/abpframework/abp/blob/22bc56e2794f0f323084df1ea680a1d1bdcded3c/templates/app/aspnet-core/test/MyCompanyName.MyProjectName.TestBase/Security/FakeCurrentPrincipalAccessor.cs

**Reply by mazarkaman** [reply]
Dear @maliming, I **have** `FakeCurrentPrincipalAccessor.cs` in my test project, but only when I declare my service with `GetRequiredService<IMainAppService> ` the `CurrentUser` has data otherwise, it's **null** e.g when I declare my service with `new` keyword.

Assuming **have** `FakeCurrentPrincipalAccessor.cs`:

- `GetRequiredService<IMainAppService>() ` -> CurrentUser has data
- `new  MainAppService(webInfoService)` -> CurrentUser is null

**Comment by maliming** [comment]
`CurrentUser` is a property of `MainAppService` which is injected by the property on `GetRequiredService`.

The `new` keyword will not set the property.
